hpx/runtime/parcelset/parcelport_impl.hpp

```diff
@@ -241,60 +242,91 @@ namespace hpx { namespace parcelset
+                    std::vector<write_handler_type>, bool
+                  , boost::shared_ptr<archive_type> const &
+                  , boost::shared_ptr<future_await_container_type> const &
+                )
+                    = &parcelport_impl::put_parcel_await;
+                (*future_await)(
+                    util::bind(
+                        util::one_shot(awaiter), this,
+                        dest, std::move(parcels), std::move(handlers), true,
+                        archive, future_await)
+                );
+                return;
+            }
+
             // enqueue the outgoing parcel ...
-            enqueue_parcel(
-                dest, std::move(p), std::move(f), std::move(future_await->new_gids_));
+            enqueue_parcels(dest, std::move(parcels), std::move(handlers),
+                std::move(future_await->new_gids_));
 
             if (trigger)
-            {
                 get_connection_and_send_parcels(dest);
-            }
         }
 
-        void put_parcels(std::vector<locality> dests, std::vector<parcel> parcels,
+        void put_parcels(locality const& dest, std::vector<parcel> parcels,
             std::vector<write_handler_type> handlers)
         {
-            if (dests.size() == parcels.size() && parcels.size() != handlers.size())
+            if (parcels.size() != handlers.size())
             {
                 HPX_THROW_EXCEPTION(bad_parameter, "parcelport::put_parcels",
                     "mismatched number of parcels and handlers");
                 return;
             }
 
-            locality const& locality_id = dests[0];
+            locality const& locality_id = dest;
 
 #if defined(HPX_DEBUG)
             // make sure all parcels go to the same locality
-            for (std::size_t i = 1; i != dests.size(); ++i)
+            HPX_ASSERT(dest.type() == type());
+            for (std::size_t i = 1; i != parcels.size(); ++i)
             {
-                HPX_ASSERT(locality_id == dests[i]);
                 HPX_ASSERT(parcels[0].destination_locality() ==
                     parcels[i].destination_locality());
             }
 #endif
 
+            boost::shared_ptr<future_await_container_type> future_await =
+                boost::make_shared<future_await_container_type>();
+            boost::shared_ptr<archive_type> archive =
+                boost::make_shared<archive_type>(*future_await);
+
             // enqueue the outgoing parcels ...
             HPX_ASSERT(parcels.size() == handlers.size());
-            for(std::size_t i = 0; i < parcels.size(); ++i)
-            {
-                put_parcel(
-                    locality_id, std::move(parcels[i]), std::move(handlers[i]),
-                    false);
-            }
+            put_parcel_await(
+                locality_id, std::move(parcels), std::move(handlers),
+                false, archive, future_await);
 
             get_connection_and_send_parcels(locality_id);
```

Comment:
we should probably check here if put_parcel_await enqueued any parcels, maybe returning `true` if it didn't had to wait for futures to avoid contention on the parcel queues.